File: backend/app/modules/marketplace/explain.py
"""`build_key_differences` para `CompareView.key_differences` (B10).

Puerto Python de `frontend/src/api/mock/engine/explain.ts::buildKeyDifferences`
(**solo lectura** en el frontend, este módulo es la implementación real del
backend): 3-4 observaciones concretas armadas **solo** a partir de
`breakdown`/`evidence_counts` ya calculados por el motor -- nunca inventa un
dato nuevo, nunca llama a IA (esto es texto derivado de números, no prosa
generada; A5 EXPLAIN sigue siendo la única pieza que redacta con un LLM,
docs/05 §7 A5).
"""

from __future__ import annotations

from app.modules.marketplace.schemas import AnonymousCandidateCard

_COMPONENT_LABELS_ES: dict[str, str] = {
    "TECHNICAL": "habilidades técnicas",
    "BEHAVIORAL": "competencias conductuales",
    "EXPERIENCE": "experiencia",
    "EVIDENCE": "calidad de evidencia",
    "SALARY": "compatibilidad salarial",
    "LOCATION": "ubicación",
}

_COMPONENTS = ("TECHNICAL", "BEHAVIORAL", "EXPERIENCE", "EVIDENCE", "SALARY", "LOCATION")


def _anon_display_code(anon_code: str) -> str:
    compact = "".join(ch for ch in anon_code if ch.isalnum())
    tail = compact[-3:].upper() if compact else ""
    return f"CANDIDATO #{tail or compact.upper()}"


def _raw_for(card: AnonymousCandidateCard, component: str) -> float:
    return next((item.raw for item in card.breakdown if item.component == component), 0.0)


def build_key_differences(cards: list[AnonymousCandidateCard]) -> list[str]:
    if len(cards) < 2:
        return []

    sorted_cards = sorted(cards, key=lambda c: c.total_score, reverse=True)
    top = sorted_cards[0]
    bottom = sorted_cards[-1]
    lines: list[str] = []

    lines.append(
        f"{_anon_display_code(top.anon_code)} lidera con {top.total_score}% de compatibilidad, "
        f"{top.total_score - bottom.total_score} puntos por encima de {_anon_display_code(bottom.anon_code)}."
    )

    tech_leader = max(cards, key=lambda c: _raw_for(c, "TECHNICAL"))
    beh_leader = max(cards, key=lambda c: _raw_for(c, "BEHAVIORAL"))
    if tech_leader.match_result_id == beh_leader.match_result_id:
        lines.append(
            f"{_anon_display_code(tech_leader.anon_code)} lidera tanto en habilidades técnicas "
            f"({round(_raw_for(tech_leader, 'TECHNICAL'))} de 100) como en competencias conductuales "
            f"({round(_raw_for(beh_leader, 'BEHAVIORAL'))} de 100)."
        )
    else:
        lines.append(
            f"{_anon_display_code(tech_leader.anon_code)} lidera en habilidades técnicas "
            f"({round(_raw_for(tech_leader, 'TECHNICAL'))} de 100), mientras que "
            f"{_anon_display_code(beh_leader.anon_code)} lidera en competencias conductuales "
            f"({round(_raw_for(beh_leader, 'BEHAVIORAL'))} de 100)."
        )

    most_verified = max(cards, key=lambda c: c.evidence_counts.get("verified", 0))
    if most_verified.evidence_counts.get("verified", 0) > 0:
        count = most_verified.evidence_counts["verified"]
        plural = "" if count == 1 else "es"
        plural2 = "" if count == 1 else "s"
        lines.append(
            f"{_anon_display_code(most_verified.anon_code)} tiene la evidencia más verificada "
            f"({count} habilidad{plural} respaldada{plural2} con documento)."
        )
    else:
        least_evaluated = min(cards, key=lambda c: c.evidence_counts.get("evaluated", 0))
        lines.append(
            f"Ningún candidato tiene habilidades verificadas todavía; {_anon_display_code(least_evaluated.anon_code)} "
            f"es quien tiene menos evidencia evaluada ({least_evaluated.evidence_counts.get('evaluated', 0)})."
        )

    best_flip: tuple[str, AnonymousCandidateCard, float] | None = None
    for challenger in sorted_cards[1:]:
        for component in _COMPONENTS:
            gap = _raw_for(challenger, component) - _raw_for(top, component)
            if gap > 0 and (best_flip is None or gap > best_flip[2]):
                best_flip = (component, challenger, gap)

    if best_flip is not None:
        component, challenger, _gap = best_flip
        lines.append(
            f"En {_COMPONENT_LABELS_ES[component]}, {_anon_display_code(challenger.anon_code)} supera a "
            f"{_anon_display_code(top.anon_code)} ({round(_raw_for(challenger, component))} vs "
            f"{round(_raw_for(top, component))} de 100); ese criterio podría cambiar la decisión final."
        )
    else:
        lines.append(
            f"{_anon_display_code(top.anon_code)} lidera en todos los criterios individuales frente al resto, "
            "no solo en el total."
        )

    return lines[:4]

File: frontend/src/api/mock/engine/explain.ts
import type { AnonymousCandidateCard, MatchComponent } from "@/api/types";
import type { StoredMatchResult } from "../state";
import { anonDisplayCode } from "../util";

/**
 * Verifica que un texto de explicación nunca mencione un porcentaje distinto
 * al total_score del candidato. Se usa en dev inmediatamente después de generar
 * el texto (ver buildExplanation) y puede reutilizarse en pruebas manuales.
 */
export function assertExplanation(text: string, totalScore: number): void {
  const matches = text.matchAll(/(\d+(?:\.\d+)?)\s*%/g);
  for (const match of matches) {
    const value = Number(match[1]);
    if (value !== totalScore) {
      throw new Error(
        `assertExplanation: el texto menciona ${value}% pero total_score es ${totalScore}%. Texto: "${text}"`,
      );
    }
  }
}

function wordLimit(text: string, max = 120): string {
  const words = text.trim().split(/\s+/);
  return words.length <= max ? text : words.slice(0, max).join(" ") + "…";
}

/**
 * Plantilla en español (≤120 palabras) a partir del breakdown ya calculado.
 * `nextScore` es el total_score del siguiente candidato en el ranking (o null
 * si es el último), usado solo para la frase de posición relativa.
 */
export function buildExplanation(card: StoredMatchResult, nextScore: number | null): string {
  const strengths = card.strengths.slice(0, 2).join(" y ");
  const gaps = card.gaps.length > 0 ? card.gaps.slice(0, 2).join(" y ") : null;
  const penaltyPoints = card.penalties.reduce((sum, p) => sum + p.points, 0);

  const displayCode = anonDisplayCode(card.anon_code);
  const positionSentence =
    nextScore != null && card.total_score > nextScore
      ? `${displayCode} queda por encima del siguiente candidato por ${card.total_score - nextScore} puntos.`
      : nextScore != null
        ? `${displayCode} está muy cerca del siguiente candidato en el ranking.`
        : `${displayCode} es el mejor evaluado de este grupo.`;

  const strengthSentence = strengths
    ? `Sus principales fortalezas evaluadas son ${strengths}.`
    : "Muestra evidencia consistente aunque sin una fortaleza destacada todavía.";

  const gapSentence = gaps
    ? `Como brecha, falta evidencia en ${gaps}.`
    : "No se detectan brechas relevantes frente a los requisitos de la vacante.";

  const penaltySentence =
    penaltyPoints > 0
      ? `Se aplicaron ${penaltyPoints} puntos de penalización por requisitos obligatorios, salario o ubicación fuera de rango.`
      : "";

  const text = wordLimit(
    `Con ${card.total_score}% de compatibilidad (${card.score_label.toLowerCase()}), ${positionSentence} ${strengthSentence} ${gapSentence} ${penaltySentence}`.trim(),
  );

  assertExplanation(text, card.total_score);
  return text;
}

/** Duplica `matchComponentLabels` de `talentLabels.ts` (el mock no puede importar de `src/features`). */
const COMPONENT_LABELS_ES: Record<MatchComponent, string> = {
  TECHNICAL: "habilidades técnicas",
  BEHAVIORAL: "competencias conductuales",
  EXPERIENCE: "experiencia",
  EVIDENCE: "calidad de evidencia",
  SALARY: "compatibilidad salarial",
  LOCATION: "ubicación",
};

function rawFor(card: AnonymousCandidateCard, component: MatchComponent): number {
  return card.breakdown.find((item) => item.component === component)?.raw ?? 0;
}

/**
 * 3-4 observaciones concretas a partir del `breakdown`/`evidence_counts` ya
 * calculados (sin inventar datos nuevos): quién lidera en general, quién en
 * técnico/conductual, quién tiene más evidencia verificada o la más limitada,
 * y en qué criterio individual el segundo lugar realmente supera al primero
 * (el dato que más podría inclinar la decisión final).
 */
export function buildKeyDifferences(cards: AnonymousCandidateCard[]): string[] {
  if (cards.length < 2) return [];
  const sorted = [...cards].sort((a, b) => b.total_score - a.total_score);
  const top = sorted[0]!;
  const bottom = sorted[sorted.length - 1]!;
  const lines: string[] = [];

  lines.push(
    `${anonDisplayCode(top.anon_code)} lidera con ${top.total_score}% de compatibilidad, ` +
      `${top.total_score - bottom.total_score} puntos por encima de ${anonDisplayCode(bottom.anon_code)}.`,
  );

  const techLeader = [...cards].sort((a, b) => rawFor(b, "TECHNICAL") - rawFor(a, "TECHNICAL"))[0]!;
  const behLeader = [...cards].sort((a, b) => rawFor(b, "BEHAVIORAL") - rawFor(a, "BEHAVIORAL"))[0]!;
  if (techLeader.match_result_id === behLeader.match_result_id) {
    lines.push(
      `${anonDisplayCode(techLeader.anon_code)} lidera tanto en habilidades técnicas ` +
        `(${Math.round(rawFor(techLeader, "TECHNICAL"))} de 100) como en competencias conductuales ` +
        `(${Math.round(rawFor(behLeader, "BEHAVIORAL"))} de 100).`,
    );
  } else {
    lines.push(
      `${anonDisplayCode(techLeader.anon_code)} lidera en habilidades técnicas ` +
        `(${Math.round(rawFor(techLeader, "TECHNICAL"))} de 100), mientras que ` +
        `${anonDisplayCode(behLeader.anon_code)} lidera en competencias conductuales ` +
        `(${Math.round(rawFor(behLeader, "BEHAVIORAL"))} de 100).`,
    );
  }

  const mostVerified = [...cards].sort((a, b) => b.evidence_counts.verified - a.evidence_counts.verified)[0]!;
  if (mostVerified.evidence_counts.verified > 0) {
    lines.push(
      `${anonDisplayCode(mostVerified.anon_code)} tiene la evidencia más verificada ` +
        `(${mostVerified.evidence_counts.verified} habilidad${mostVerified.evidence_counts.verified === 1 ? "" : "es"} respaldada${mostVerified.evidence_counts.verified === 1 ? "" : "s"} con documento).`,
    );
  } else {
    const leastEvaluated = [...cards].sort(
      (a, b) => a.evidence_counts.evaluated - b.evidence_counts.evaluated,
    )[0]!;
    lines.push(
      `Ningún candidato tiene habilidades verificadas todavía; ${anonDisplayCode(leastEvaluated.anon_code)} ` +
        `es quien tiene menos evidencia evaluada (${leastEvaluated.evidence_counts.evaluated}).`,
    );
  }

  const components: MatchComponent[] = ["TECHNICAL", "BEHAVIORAL", "EXPERIENCE", "EVIDENCE", "SALARY", "LOCATION"];
  let bestFlip: { component: MatchComponent; challenger: AnonymousCandidateCard; gap: number } | null = null;
  for (const challenger of sorted.slice(1)) {
    for (const component of components) {
      const gap = rawFor(challenger, component) - rawFor(top, component);
      if (gap > 0 && (bestFlip == null || gap > bestFlip.gap)) {
        bestFlip = { component, challenger, gap };
      }
    }
  }
  if (bestFlip) {
    lines.push(
      `En ${COMPONENT_LABELS_ES[bestFlip.component]}, ${anonDisplayCode(bestFlip.challenger.anon_code)} supera a ` +
        `${anonDisplayCode(top.anon_code)} (${Math.round(rawFor(bestFlip.challenger, bestFlip.component))} vs ` +
        `${Math.round(rawFor(top, bestFlip.component))} de 100); ese criterio podría cambiar la decisión final.`,
    );
  } else {
    lines.push(
      `${anonDisplayCode(top.anon_code)} lidera en todos los criterios individuales frente al resto, no solo en el total.`,
    );
  }

  return lines.slice(0, 4);
}
